io_buffer.c

```diff
@@ -2120,6 +2155,363 @@ io_buffer_pwrite(VALUE self, VALUE io, VALUE length, VALUE offset)
+{
+    for (size_t offset = 0; offset < size; offset += 1) {
+      output[offset] = base[offset] & mask[offset % mask_size];
+    }
+}
+
+/*
+ *  call-seq:
+ *    source & mask -> io_buffer
+ *
+ *  Generate a new buffer the same size as the source by applying the binary AND
+ *  operation to the source, using the mask, repeating as necessary.
+ *
+ *    IO::Buffer.for("1234567890") & IO::Buffer.for("\xFF\x00\x00\xFF")
+ *    # =>
+ *    # #<IO::Buffer 0x00005589b2758480+4 INTERNAL>
+ *    # 0x00000000  31 00 00 34 35 00 00 38 39 00                   1..45..89.
+ */
+static VALUE
+io_buffer_and(VALUE self, VALUE mask)
+{
+    struct rb_io_buffer *data = NULL;
+    TypedData_Get_Struct(self, struct rb_io_buffer, &rb_io_buffer_type, data);
+
+    struct rb_io_buffer *mask_data = NULL;
+    TypedData_Get_Struct(mask, struct rb_io_buffer, &rb_io_buffer_type, mask_data);
+
+    io_buffer_check_mask(mask_data);
+
+    VALUE output = rb_io_buffer_new(NULL, data->size, io_flags_for_size(data->size));
+    struct rb_io_buffer *output_data = NULL;
+    TypedData_Get_Struct(output, struct rb_io_buffer, &rb_io_buffer_type, output_data);
+
+    memory_and(output_data->base, data->base, data->size, mask_data->base, mask_data->size);
+
+    return output;
+}
+
+static void
+memory_or(unsigned char * restrict output, unsigned char * restrict base, size_t size, unsigned char * restrict mask, size_t mask_size)
+{
+    for (size_t offset = 0; offset < size; offset += 1) {
+      output[offset] = base[offset] | mask[offset % mask_size];
+    }
+}
+
+/*
+ *  call-seq:
+ *    source | mask -> io_buffer
+ *
+ *  Generate a new buffer the same size as the source by applying the binary OR
+ *  operation to the source, using the mask, repeating as necessary.
+ *
+ *    IO::Buffer.for("1234567890") | IO::Buffer.for("\xFF\x00\x00\xFF")
+ *    # =>
+ *    # #<IO::Buffer 0x0000561785ae3480+10 INTERNAL>
+ *    # 0x00000000  ff 32 33 ff ff 36 37 ff ff 30                   .23..67..0
+ */
+static VALUE
+io_buffer_or(VALUE self, VALUE mask)
+{
+    struct rb_io_buffer *data = NULL;
+    TypedData_Get_Struct(self, struct rb_io_buffer, &rb_io_buffer_type, data);
+
+    struct rb_io_buffer *mask_data = NULL;
+    TypedData_Get_Struct(mask, struct rb_io_buffer, &rb_io_buffer_type, mask_data);
+
+    io_buffer_check_mask(mask_data);
+
+    VALUE output = rb_io_buffer_new(NULL, data->size, io_flags_for_size(data->size));
+    struct rb_io_buffer *output_data = NULL;
+    TypedData_Get_Struct(output, struct rb_io_buffer, &rb_io_buffer_type, output_data);
+
+    memory_or(output_data->base, data->base, data->size, mask_data->base, mask_data->size);
+
+    return output;
+}
+
+static void
+memory_xor(unsigned char * restrict output, unsigned char * restrict base, size_t size, unsigned char * restrict mask, size_t mask_size)
+{
+    for (size_t offset = 0; offset < size; offset += 1) {
+      output[offset] = base[offset] ^ mask[offset % mask_size];
```

Comment:
`% mask_size` is likely quite slower than a bit mask (if the size is power of two), or condition with a separate mask_index resetting to 0 if >= mask_size.

Reply:
I checked the assembly and it looks to generate fairly efficient SSE/AVX instructions, so I trust the compiler is doing the right things here. In any case, simple and correct is a good first implementation, and later on we can unroll it by hand if performance is an issue.

Reply:
It's not about unrolling. It's about a modulo operation is as expensive as division (many cycles), and a lot more than `+ - * & ! ^`.

Reply:
I played around with this: https://godbolt.org/z/7MEfsrz9z - change -Os to -O3 and you see the compiler will unroll the loop and load the data in chunks. My understanding is that by unrolling the loop carefully, some division operations can be skipped.

If you know how we can count the number of divs being performed, that would be awesome, otherwise, I'm fine with slow but correct to start with, and figuring out a faster implementation later as an improvement. If you think adding a branch to the inner loop will ultimately be faster, let's do that? But at that point I think we need to benchmark it.

Reply:
I'm fairly confident it does not remove the division operations, it could only do that if it did the branch I suggest, and there are plenty of `divl` and `divq` left in that assembly.
Maybe it does a few less, but it could only do that without branches if it can make sure the mask_size is e.g. > 1, and it cannot know that.

I agree, we should benchmark it, then it should be clear. With the explicit condition there would be 0 division left, but the extra branch. It should still unroll fine.
The best would be if we could have the `mask_size` as a constant in the source, this should be possible by doing an early check and call with a constant for the `mask_size` we care about + force inline this function.